File: Parcial 01/parcial01.py
import psycopg2
import re

#CONEXIÓN A BASE DE DATOS

def conectar_bd():
    try:
        conn = psycopg2.connect(
            "dbname=gasolinera user=postgres password=koala host=localhost port=5432"
        )
        return conn
    except Exception as e:
        print("Error al conectar con la base de datos:", e)
        return None

# CREAR TABLA EN BASE DE DATOS

def inicializar_bd():
    """Crea la tabla en la base de datos si no existe"""
    try:
        conn = conectar_bd()
        if conn:
            with conn.cursor() as cur:
                cur.execute("""
                    CREATE TABLE IF NOT EXISTS facturas_gasolina (
                        cliente TEXT NOT NULL,
                        placa TEXT NOT NULL,
                        combustible TEXT NOT NULL,
                        litros REAL NOT NULL,
                        total REAL NOT NULL
                    )
                """)
                conn.commit()
            conn.close()
    except Exception as e:
        print("Error inicializando base de datos:", e)



# VALIDACIONES DE FORMATO

def pedir_nombre_apellido(mensaje):
    patron = r"^[A-Za-z]+(?: [A-Za-z]+)+$"
    while True:
        nombre = input(mensaje).strip()
        if re.match(patron, nombre):
            return nombre
        else:
            print("Ingrese nombre y apellido(s) separados por espacio. Solo letras sin signos ni números.")


def pedir_placa(mensaje):
    patron = r"^P[0-9]{3}[A-Z]{3}$"
    while True:
        placa = input(mensaje).strip().upper()
        if re.match(patron, placa):
            return placa
        else:
            print("La placa debe tener formato P123ABC.")


def pedir_entero(mensaje, minimo=1, maximo=None):
    while True:
        try:
            valor = int(input(mensaje))
            if valor < minimo:
                print(f"El valor debe ser mayor o igual a {minimo}")
            elif maximo and valor > maximo:
                print(f"El valor debe ser menor o igual a {maximo}")
            else:
                return valor
        except ValueError:
            print("Debe ingresar un número entero válido.")


def pedir_flotante(mensaje, minimo=0.1):
    while True:
        try:
            valor = float(input(mensaje))
            if valor < minimo:
                print(f"El valor debe ser mayor a {minimo}")
            else:
                return valor
        except ValueError:
            print("Debe ingresar un número válido.")



# LÓGICA DEL PROGRAMA

precios_combustible = {
    1: ("Gasolina Regular", 7.84),
    2: ("Gasolina Premium", 8.23),
    3: ("Diesel", 7.39)
}


def calcular_total(tipo, litros):
    nombre, precio = precios_combustible[tipo]
    return nombre, litros * precio


def guardar_factura(cliente, placa, combustible, litros, total):
    try:
        # Guardar en archivo
        with open("salida.txt", "a", encoding="utf-8") as f:
            f.write(f"{cliente},{placa},{combustible},{litros},{total:.2f}\n")

        # Guardar en BD
        conn = conectar_bd()
        if conn:
            with conn.cursor() as cur:
                cur.execute(
                    "INSERT INTO facturas_gasolina (cliente, placa, combustible, litros, total) VALUES (%s,%s,%s,%s,%s)",
                    (cliente, placa, combustible, litros, total)
                )
                conn.commit()
            conn.close()
        print("Factura guardada correctamente.")
    except Exception as e:
        print("Error al guardar factura:", e)


def mostrar_historial():
    print("\n--- Historial desde archivo ---")
    try:
        with open("salida.txt", "r", encoding="utf-8") as f:
            for linea in f:
                print(" -", linea.strip())
    except FileNotFoundError:
        print("No hay facturas registradas en archivo.")

    print("\n--- Historial desde base de datos ---")
    try:
        conn = conectar_bd()
        if conn:
            with conn.cursor() as cur:
                cur.execute("SELECT cliente, placa, combustible, litros, total FROM facturas_gasolina")
                registros = cur.fetchall()
                if registros:
                    for r in registros:
                        print(f"Cliente={r[0]}, Placa={r[1]}, Combustible={r[2]}, Litros={r[3]}, Total=Q{r[4]:.2f}")
                else:
                    print("No hay facturas en la base de datos.")
            conn.close()
    except Exception as e:
        print("Error mostrando historial:", e)


def borrar_historial():
    print("\n--- BORRAR HISTORIAL ---")
    print("1. Borrar una factura específica")
    print("2. Borrar todas las facturas")
    print("3. Borrar todas las facturas de una placa")
    print("4. Cancelar")

    opcion = pedir_entero("Seleccione una opción: ", 1, 4)

    if opcion == 1:
        placa = input("Ingrese la placa del vehículo de la factura a borrar: ").strip().upper()

        # Borrar en archivo
        try:
            nuevas_lineas = []
            with open("salida.txt", "r", encoding="utf-8") as f:
                for linea in f:
                    if not (placa in linea):
                        nuevas_lineas.append(linea)
            with open("salida.txt", "w", encoding="utf-8") as f:
                f.writelines(nuevas_lineas)
            print("Factura borrada del archivo (si existía).")
        except FileNotFoundError:
            print("No hay archivo de facturas.")

        # Borrar en BD
        try:
            conn = conectar_bd()
            if conn:
                with conn.cursor() as cur:
                    cur.execute("DELETE FROM facturas_gasolina WHERE placa=%s", (placa,))
                    conn.commit()
                conn.close()
                print("Factura borrada de la base de datos (si existía).")
        except Exception as e:
            print("Error borrando factura en base de datos:", e)

    elif opcion == 2:
        # BORRAR TODO
        try:
            with open("salida.txt", "w", encoding="utf-8"):
                pass
            print("Archivo borrado correctamente.")
        except Exception as e:
            print("Error borrando archivo:", e)

        try:
            conn = conectar_bd()
            if conn:
                with conn.cursor() as cur:
                    cur.execute("DELETE FROM facturas_gasolina")
                    conn.commit()
                conn.close()
                print("Historial en base de datos borrado.")
        except Exception as e:
            print("Error borrando historial en base de datos:", e)

    elif opcion == 3:
        placa = input("Ingrese la placa de las facturas a borrar: ").strip().upper()

        # Borrar en archivo
        try:
            nuevas_lineas = []
            with open("salida.txt", "r", encoding="utf-8") as f:
                for linea in f:
                    if placa not in linea:
                        nuevas_lineas.append(linea)
            with open("salida.txt", "w", encoding="utf-8") as f:
                f.writelines(nuevas_lineas)
            print(f"Facturas con placa {placa} borradas del archivo (si existían).")
        except FileNotFoundError:
            print("No hay archivo de facturas.")

        # Borrar en BD
        try:
            conn = conectar_bd()
            if conn:
                with conn.cursor() as cur:
                    cur.execute("DELETE FROM facturas_gasolina WHERE placa=%s", (placa,))
                    conn.commit()
                conn.close()
                print(f"Facturas con placa {placa} borradas de la base de datos (si existían).")
        except Exception as e:
            print("Error borrando facturas en base de datos:", e)

    elif opcion == 4:
        print("Cancelando borrado. Regresando al menú principal...")



# FUNCIÓN PRINCIPAL

def menu():
    inicializar_bd()
    while True:
        print("\n========== MENÚ PRINCIPAL ==========")
        print("1. Registrar factura")
        print("2. Ver historial")
        print("3. Borrar historial")
        print("4. Salir")

        opcion = pedir_entero("Seleccione una opción (1-4): ", 1, 4)

        if opcion == 1:
            cliente = pedir_nombre_apellido("Nombre y apellido del cliente: ")
            placa = pedir_placa("Placa del vehículo (P123ABC): ")

            print("\n-----Tipos de combustible-----")
            for k, v in precios_combustible.items():
                print(f"{k}. {v[0]} - Q{v[1]:.2f} por litro")

            tipo = pedir_entero("Seleccione el tipo (1-3): ", 1, 3)
            litros = pedir_flotante("Litros a despachar: ", 0.1)

            combustible, total = calcular_total(tipo, litros)

            print("\n--- FACTURA ---")
            print(f"Cliente: {cliente}")
            print(f"Vehículo: {placa}")
            print(f"Combustible: {combustible}")
            print(f"Litros: {litros:.2f}")
            print(f"Total a pagar: Q{total:.2f}")
            print("---------------------")

            guardar_factura(cliente, placa, combustible, litros, total)

        elif opcion == 2:
            mostrar_historial()

        elif opcion == 3:
            borrar_historial()

        elif opcion == 4:
            print("Gracias por usar el sistema. Saliendo...")
            break

# EJECUCIÓN

menu()

File: Parcial 01/parcial01.m
#Facturación Gasolinera

# Comprobar si es Octave
if (exist('OCTAVE_VERSION','builtin') ~= 0)
  pkg load signal;
  pkg load database;
end

# Ruta del archivo de facturas
archivo_facturas = "C:\\Users\\Alejandra\\Desktop\\980-Proyectos\\Parcial 01\\salida.txt";

# Conexión a la base de datos
conn = pq_connect(setdbopts( ...
    'dbname','gasolinera', ...
    'host','localhost', ...
    'port','5432', ...
    'user','postgres', ...
    'password','koala'));

# Crear tabla si no existe
create_table = [ ...
"CREATE TABLE IF NOT EXISTS facturas_gasolina (", ...
" cliente TEXT NOT NULL,", ...
" placa TEXT NOT NULL,", ...
" combustible TEXT NOT NULL,", ...
" litros REAL NOT NULL,", ...
" total REAL NOT NULL );" ];

pq_exec_params(conn, create_table);

#Funciones auxiliares

function nombre = pedir_nombre()
  valido = false;
  while ~valido
    nombre = strtrim(input("Ingrese nombre y apellido: ", "s"));
    if ~isempty(regexp(nombre, "^[A-Za-z]+( [A-Za-z]+)+$", "once"))
      valido = true;
    else
      disp("Formato inválido. Ejemplo: Juan Pérez");
    endif
  endwhile
endfunction

function placa = pedir_placa()
  valido = false;
  while ~valido
    placa = upper(strtrim(input("Ingrese placa (P123ABC): ", "s")));
    if ~isempty(regexp(placa, "^P[0-9]{3}[A-Z]{3}$", "once"))
      valido = true;
    else
      disp("Formato inválido. Ejemplo: P123ABC");
    endif
  endwhile
endfunction

function valor = pedir_entero(msg, minval, maxval)
  valido = false;
  while ~valido
    valor = str2num(input(msg, "s"));
    if isempty(valor) || ~isscalar(valor) || valor < minval || valor > maxval
      printf("Debe ingresar un entero válido entre %d y %d\n", minval, maxval);
    else
      valido = true;
    endif
  endwhile
endfunction

function valor = pedir_flotante(msg, minval)
  valido = false;
  while ~valido
    valor = str2double(input(msg, "s"));
    if isnan(valor) || valor < minval
      printf("Debe ingresar un número mayor que %.2f\n", minval);
    else
      valido = true;
    endif
  endwhile
endfunction

#Logica del programa

precios = {
  1, "Gasolina Regular", 7.84;
  2, "Gasolina Premium", 8.23;
  3, "Diesel", 7.39
};

function guardar_factura(conn, archivo_facturas, cliente, placa, combustible, litros, total)
  # Guardar en archivo
  fid = fopen(archivo_facturas, "a");
  fprintf(fid, "%s,%s,%s,%.2f,%.2f\n", cliente, placa, combustible, litros, total);
  fclose(fid);

  # Guardar en BD
  query = "INSERT INTO facturas_gasolina (cliente, placa, combustible, litros, total) VALUES ($1,$2,$3,$4,$5)";
  pq_exec_params(conn, query, {cliente, placa, combustible, litros, total});
  disp("Factura guardada correctamente.");
endfunction

function mostrar_historial(archivo_facturas)
  disp("\n--- HISTORIAL desde archivo ---");
  fid = fopen(archivo_facturas, "r");
  if fid != -1
    while !feof(fid)
      linea = fgetl(fid);
      if ischar(linea)
        disp([" - " linea]);
      endif
    endwhile
    fclose(fid);
  else
    disp("No hay facturas registradas en archivo.");
  endif
endfunction

function borrar_historial(conn, archivo_facturas)
  disp("\n--- BORRAR HISTORIAL ---");
  disp("1. Borrar una factura específica (por placa)");
  disp("2. Borrar todas las facturas");
  disp("3. Borrar todas las facturas de una placa");
  disp("4. Cancelar");

  opcion = pedir_entero("Seleccione una opción: ", 1, 4);

  if opcion == 1
    placa = pedir_placa();

    # borrar en archivo (solo la primera coincidencia)
    if exist(archivo_facturas, "file")
      fid = fopen(archivo_facturas, "r");
      lineas = {};
      eliminado = false;
      while !feof(fid)
        linea = fgetl(fid);
        if ischar(linea)
          if ~eliminado && ~isempty(strfind(linea, placa))
            eliminado = true; # omitimos esta línea
          else
            lineas{end+1} = linea;
          endif
        endif
      endwhile
      fclose(fid);
      fid = fopen(archivo_facturas, "w");
      for i=1:length(lineas)
        fprintf(fid, "%s\n", lineas{i});
      endfor
      fclose(fid);
      if eliminado
        printf("Factura con placa %s borrada del archivo.\n", placa);
      else
        printf("No se encontró factura con placa %s en el archivo.\n", placa);
      endif
    endif

    # borrar en BD (solo una fila usando CTE)
    query = ["WITH cte AS (SELECT ctid FROM facturas_gasolina WHERE placa=$1 LIMIT 1) ", ...
             "DELETE FROM facturas_gasolina WHERE ctid IN (SELECT ctid FROM cte)"];
    pq_exec_params(conn, query, {placa});
    disp("Factura borrada de la base de datos (si existía).");

  elseif opcion == 2
    # borrar todo
    fid = fopen(archivo_facturas, "w"); fclose(fid);
    disp("Archivo borrado correctamente.");
    pq_exec_params(conn, "DELETE FROM facturas_gasolina");
    disp("Historial en base de datos borrado.");

  elseif opcion == 3
    placa = pedir_placa();
    # borrar todas las facturas de esa placa en archivo
    if exist(archivo_facturas, "file")
      fid = fopen(archivo_facturas, "r");
      lineas = {};
      while !feof(fid)
        linea = fgetl(fid);
        if ischar(linea) && isempty(strfind(linea, placa))
          lineas{end+1} = linea;
        endif
      endwhile
      fclose(fid);
      fid = fopen(archivo_facturas, "w");
      for i=1:length(lineas)
        fprintf(fid, "%s\n", lineas{i});
      endfor
      fclose(fid);
      printf("Facturas con placa %s borradas del archivo (si existían).\n", placa);
    endif
    # borrar en BD
    pq_exec_params(conn, "DELETE FROM facturas_gasolina WHERE placa=$1", {placa});
    printf("Facturas con placa %s borradas de la base de datos (si existían).\n", placa);

  else
    disp("Cancelando borrado. Regresando al menú principal...");
  endif
endfunction

#MENU PRINCIPAL

while true
  disp("\n========== MENÚ PRINCIPAL ==========");
  disp("1. Registrar factura");
  disp("2. Ver historial");
  disp("3. Borrar historial");
  disp("4. Salir");

  opcion = pedir_entero("Seleccione una opción (1-4): ", 1, 4);

  if opcion == 1
    cliente = pedir_nombre();
    placa = pedir_placa();

    disp("\n----- Tipos de combustible -----");
    for i=1:rows(precios)
      printf("%d. %s - Q%.2f por litro\n", precios{i,1}, precios{i,2}, precios{i,3});
    endfor

    tipo = pedir_entero("Seleccione el tipo (1-3): ", 1, 3);
    litros = pedir_flotante("Litros a despachar: ", 0.1);

    combustible = precios{tipo,2};
    total = litros * precios{tipo,3};

    disp("\n--- FACTURA ---");
    printf("Cliente: %s\n", cliente);
    printf("Vehículo: %s\n", placa);
    printf("Combustible: %s\n", combustible);
    printf("Litros: %.2f\n", litros);
    printf("Total a pagar: Q%.2f\n", total);
    disp("---------------------");

    guardar_factura(conn, archivo_facturas, cliente, placa, combustible, litros, total);

  elseif opcion == 2
    mostrar_historial(archivo_facturas);

  elseif opcion == 3
    borrar_historial(conn, archivo_facturas);

  elseif opcion == 4
    disp("Gracias por usar el sistema. Saliendo...");
    break;
  endif
endwhile

# Cerrar conexión
pq_close(conn);

